kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1288 @@
+        }
+
+        // Gets the array of conditions from the modeler
+        ConditionsArrayType& r_conditions_array = rModelPart.Conditions();
+        const int number_of_conditions = static_cast<int>(r_conditions_array.size());
+        #pragma omp for  schedule(guided, 512)
+        for (int i = 0; i < number_of_conditions; ++i) {
+            auto it_cond = r_conditions_array.begin() + i;
+
+            // Gets list of Dof involved on every element
+            pScheme->GetConditionDofList(*(it_cond.base()), dof_list, r_current_process_info);
+            dofs_aux_list_solvable.insert(dof_list.begin(), dof_list.end());
+        }
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", ( this->GetEchoLevel() > 2)) << "Initializing constraints loop" << std::endl;
+
+        // We don't do a loop in OMP because erase is not a parallel threadsafe operation
+        auto& r_constraints_array = rModelPart.MasterSlaveConstraints();
+        const int number_of_constraints = static_cast<int>(r_constraints_array.size());
+
+        for (int i = 0; i < number_of_constraints; ++i) {
+            auto it_const = r_constraints_array.begin() + i;
+
+            // Gets list of Dof involved on every element
+            it_const->GetDofList(dof_list, auxiliar_dof_list, r_current_process_info);
+            // We remove the slave dofs
+            for (auto& dof : dof_list) {
+                auto it = dofs_aux_list_solvable.find(dof);
+                // Check if Iterator is valid
+                if(it != dofs_aux_list_solvable.end())
+                    dofs_aux_list_solvable.erase(it);
+            }
+            dofs_aux_list_solvable.insert(auxiliar_dof_list.begin(), auxiliar_dof_list.end()); // We add the master dofs
+        }
+
+        dof_temp_solvable.reserve(dofs_aux_list_solvable.size());
+        for (auto& dof : dofs_aux_list_solvable) {
+            dof_temp_solvable.push_back( dof.get() );
+        }
+        dof_temp_solvable.Sort();
+        mDoFToSolveSet = dof_temp_solvable;
+
+        // Throws an exception if there are no Degrees Of Freedom involved in the analysis
+        KRATOS_ERROR_IF(BaseType::mDofSet.size() == 0) << "No degrees of freedom!" << std::endl;
+        KRATOS_ERROR_IF( mDoFToSolveSet.size() == 0) << "No degrees of freedom to solve!" << std::endl;
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", ( this->GetEchoLevel() > 2)) << "Number of degrees of freedom:" << BaseType::mDofSet.size() << std::endl;
+
+        BaseType::mDofSetIsInitialized = true;
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", ( this->GetEchoLevel() > 2 && rModelPart.GetCommunicator().MyPID() == 0)) << "Finished setting up the dofs" << std::endl;
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", ( this->GetEchoLevel() > 2)) << "Initializing lock array" << std::endl;
+
+#ifdef _OPENMP
```

Comment:
i would seriously get rid of these array of locks...atomics are faster!!

Reply:
OK